timemachine/md/minimizer.py

```diff
@@ -172,15 +183,94 @@ def minimize_host_4d(mols, host_system, host_coords, ff, box, mol_coords=None) -
-        if not np.all(norm < MAX_FORCE_NORM):
-            raise MinimizationError(
-                f"Minimization failed to reduce large forces below threshold: |frc| = {norm} > {MAX_FORCE_NORM}"
-            )
+        check_force_norm(du_dx)
 
     return final_coords[:num_host_atoms]
 
 
+def make_gpu_impl(potentials):
+    """return bound impl of a SummedPotential constructed from potentials"""
+
+    params = [p.params for p in potentials]
+    flat_params = np.concatenate([param.reshape(-1) for param in params])
+
+    generic_potentials = [generic.from_gpu(p) for p in potentials]
+    summed_potential = generic.SummedPotential(generic_potentials, params)
+
+    return summed_potential.to_gpu().bind(flat_params).bound_impl(np.float32)
+
+
+def make_host_du_dx_fxn(mols, host_system, host_coords, ff, box, mol_coords=None):
+    """construct function to compute du_dx w.r.t. host coords, given fixed mols and box"""
+
+    # openmm host_system -> timemachine host_bps
+    host_bps, _ = openmm_deserializer.deserialize_system(host_system, cutoff=1.2)
+
+    # construct appropriate topology from (mols, ff)
+    if len(mols) == 1:
+        top = topology.BaseTopology(mols[0], ff)
+    elif len(mols) == 2:
+        top = topology.DualTopologyMinimization(mols[0], mols[1], ff)
+    else:
+        raise ValueError("mols must be length 1 or 2")
+
+    hgt = topology.HostGuestTopology(host_bps, top)
+
+    # bound impls of potentials @ lam=0 (fully coupled) endstate
+    params_potential_pairs = parameterize_system(hgt, ff, 0.0)
+    bound_potentials = [potential.bind(params) for params, potential in params_potential_pairs]
+    gpu_impl = make_gpu_impl(bound_potentials)
+
+    # read conformers from mol_coords if given, or each mol's conf0 otherwise
+    conf_list = [np.array(host_coords)]
+
+    if mol_coords is not None:
+        for mc in mol_coords:
+            conf_list.append(mc)
+    else:
+        for mol in mols:
+            conf_list.append(get_romol_conf(mol))
+
+    combined_coords = np.concatenate(conf_list)
+
+    # wrap gpu_impl, partially applying box, mol coords
+    num_host_atoms = host_coords.shape[0]
+    assert box.shape == (3, 3)
+
+    def du_dx_host_fxn(x_host):
+        x = np.array(combined_coords)
+        x[:num_host_atoms] = x_host
+
+        du_dx, _ = gpu_impl.execute(x, box)
+        du_dx_host = du_dx[:num_host_atoms]
+        return du_dx_host
+
+    return du_dx_host_fxn
+
+
+def equilibrate_host_barker(
+    mols, host_system, host_coords, ff, box, mol_coords=None, temperature=300.0, proposal_stddev=0.0001, n_steps=1000
+):
+    """possibly simpler alternative to minimize_host_4d, for purposes of
+    clash resolution and initial pre-equilibration"""
+
+    du_dx_host_fxn = make_host_du_dx_fxn(mols, host_system, host_coords, ff, box, mol_coords)
+    grad_log_q = lambda x_host: -du_dx_host_fxn(x_host) / (BOLTZ * temperature)
+
+    # TODO: if needed, revisit choice to omit Metropolis correction
```

Comment:
Can you comment on why it seems valid to ignore the Metropolis correction? Do we expect this to approximately correctly sample from the finite temperature distribution without the correction?

Reply:
>Can you comment on why it seems valid to ignore the Metropolis correction? Do we expect this to approximately correctly sample from the finite temperature distribution without the correction?

In the limit of `proposal_stddev` approaching 0, I think this proposal process should sample from the finite temperature distribution even without Metropolis correction. (Following the "local balancing" argument in the reference, and based on numerical experiments.)

Omitting the Metropolis correction when `proposal_stddev > 0` is not something that I think was intended by the authors, and I don't think it's mentioned / recommended in the paper.

Anecdotally, I would not expect this to outperform our default approximate sampler that omits a Metropolis correction (BAOAB) once equilibrated.

(I would expect the sampling bias would be worse than BAOAB's at a comparable "step size". I'd also expect this move to be much less productive per gradient evaluation, since it's not inertial -- similar to how BAOAB with `friction=+np.inf` is less efficient than BAOAB with `friction=1.0`.)

The main advantage is robustness w.r.t. clashy initialization.

Anecdotally, the unadjusted Barker proposal with proposal_stddev = 0.0001 nm appears to be good enough to resolve Lennard-Jones clashes, and can make progress even when `|force| = +inf`.

...

I'll flesh out some comments in the docstring, add optional argument to use Metropolis correction, and add some other guardrails now that `proposal_stddev` is exposed to a user of `host_equilibrate_barker`.

Reply:
Sounds good. For this PR I think just a note in the docstring would be sufficient.

Reply:
Added these notes, and modified test to exercise the case of temperature == 0 (aka grad_log_q is all +/-infs) https://github.com/proteneer/timemachine/pull/978/commits/38e76d141b789103e78d4eeb7e92f0e6902aa506

Did not add metropolis correction, but added assertion preventing user from setting proposal_stddev larger than max tested value.